### DISCUSSION TITLE
Float and boolean in JSON as strings, for later use with metadata.map?

### DISCUSSION THREAD

**Discussion by Moonbase59** [savonet/liquidsoap#3742] [Q&A]
For an external process that defines `liq_cue_in`, `liq_cue_out`, `liq_cross_duration` and others, that returns its calculations as JSON and shall be used with `metadata.map`:

**Should the numeric and boolean JSON values be coded as such, or as strings?**

I’m asking because I look at the metadata generated by an annotated playlist, and it looks like everything is a string:

```
2024/02/14 08:34:34 [m:3] ("duration", "376.05")
2024/02/14 08:34:34 [m:3] ("liq_amplify", "-7.60 dB")
2024/02/14 08:34:34 [m:3] ("liq_cross_duration", "11.60")
2024/02/14 08:34:34 [m:3] ("liq_cue_in", "0.30")
2024/02/14 08:34:34 [m:3] ("liq_cue_out", "372.00")
2024/02/14 08:34:34 [m:3] ("longtail", "false")
2024/02/14 08:34:34 [m:3] ("replaygain_album_gain", "-6.92 dB")
2024/02/14 08:34:34 [m:3] ("replaygain_album_peak", "1.126437")
2024/02/14 08:34:34 [m:3] ("replaygain_album_range", "4.71 dB")
2024/02/14 08:34:34 [m:3] ("replaygain_reference_loudness", "-18.00 LUFS")
2024/02/14 08:34:34 [m:3] ("replaygain_track_gain", "-7.58 dB")
2024/02/14 08:34:34 [m:3] ("replaygain_track_peak", "1.021502")
2024/02/14 08:34:34 [m:3] ("replaygain_track_range", "4.23 dB")
```

Does `metadata.map` expect everything as a string?

Looks to me like list of 2-tuples of type string…

**Comment by toots** [comment]
Metadata are per definition list of pairs of strings: `[(string * string)]`. I'm not sure how that relates to JSON.

If you're looking at a decoding pipeline from JSON to metadata, for instance something like parsing `{ liq_cross_duration: 11.60, liq_cue_in: 0.30}`, you can either parse with an explicit type and manually convert to string:
```liquidsoap
let json.parse ({ liq_cross_duration, liq_cue_in}:{liq_cross_duration,  liq_cue_in}) = json_string
[ ("liq_cross_duration", "#{liq_cross_duration}"), (" liq_cue_in", "#{ liq_cue_in}") ]
```


You can also parse a an associative list but you'll only get values of the same type within the JSON object. Then, you can map all values at once:
```liquidsoap
let json.parse (l: [(string * float)] as json.object) = json_string
list.map(fun (e) -> (fst(e), "#{snd(e)}"), l)
```

Lastly, you can be intrepid and use the generic parser without any type coercicion. This might work before we know how to convert any value to string:
```liquidsoap
let json.parse l = json_string
list.map(fun (e) -> (fst(e), "#{snd(e)}"), l)
```

None of this has been tested, this is just some quick ideas :-)

**Reply by Moonbase59** [reply]
Thanks for the help! I _was_ thinking of using an assoc list but didn’t get the syntax right. And the below (which works) looks _so_ darn clumsy:

```
def get_data(m) =
  log(label="get_data", 'Getting data for "' ^ m["filename"] ^ '" ...')
  
  result = list.hd(default="",
    process.read.lines(process.quote.command("cue_file", args=[m["filename"]])))
  log(label="get_data", result)
  
  let json.parse ({
    duration,
    liq_cue_in,
    liq_cue_out,
    longtail,
    liq_cross_duration,
    loudness,
    liq_amplify
  } : {
    duration: string,
    liq_cue_in: string,
    liq_cue_out: string,
    longtail: string,
    liq_cross_duration: string,
    loudness: string,
    liq_amplify: string
  }) = result
  
  [
    ("liq_test", "get_data"),
    ("duration", duration),
    ("liq_cue_in", liq_cue_in),
    ("liq_cue_out", liq_cue_out),
    ("longtail", longtail),
    ("liq_cross_duration", liq_cross_duration),
    ("loudness", loudness),
    ("liq_amplify", liq_amplify)
  ]
end

radio = metadata.map(get_data, radio)
```

The JSON response from my script (now) looks like:
```json
{"duration": "237.45", "liq_cue_in": "0.30", "liq_cue_out": "235.20", "longtail": "false", "liq_cross_duration": "1.80", "loudness": "-14.20 dB", "liq_amplify": "-3.80 dB"}
```
